### ISSUE TITLE
Unicode combining diacritical marks extracted separately

### ISSUE THREAD

**Issue by asbjornu** [UglyToad/PdfPig#439]
For some PDF documents, [combining diacritical marks](https://en.wikipedia.org/wiki/Combining_Diacritical_Marks) such as [diaeresis (¨)](https://en.wikipedia.org/wiki/Diaeresis_(diacritic)) seems to confuse the text extraction in PdfPig. In some occurrences, regular hyphens also seem to introduce an unwanted space. Both `Page.GetWords()` and `ContentOrderTextExtractor.GetText()` seem confused, as illutrated in the table below:

<table>
<thead>
<tr><th>Source</th><th>Result</th><th>Extraction method</th></tr>
</thead>
<tbody>
<tr>
<td rowspan="2"><code>Tjänsteinformation Foobar</code></td>
<td><code>Tjansteinformation Foobar ̈</code></td>
<td><code>Page.GetWords()</code></td>
</tr>
<tr>
<td><code>Tjä nsteinformation Foobar</code></td>
<td rowspan="2"><code>ContentOrderTextExtractor.GetText()</code></td>
</tr>
<tr>
<td><code>Foo-Bar</code></td>
<td><code>Foo- Bar</code></td>
</tr>
</tbody>
</table>

I have a PDF I can share that demonstrate these problems, but I'm unable to share it publicly.

**Comment by asbjornu** [comment]
If I could get any pointers to where I should start looking to fix this bug, I would be up for trying to fix this myself, @EliotJones.

**Comment by EliotJones** [comment]
So I started looking into it yesterday but didn't get very far, so it seems there's a bunch of standalone diacritics separate to those used with the combining character. There's a few listed here but I found a list of about 12 elsewhere but since lost it.  https://stackoverflow.com/questions/69847607/use-icu-to-convert-combining-diacritics-to-non-combining-equivalent

I initially considered combining the diacritics with the letters in the content stream processor logic but I think it's easier to preserve them as separate letters and have logic in GetWords to combine them in the resulting word text.

That is:

```
word: {
   text: "ö",
   letters: [
       { text: "o"},
       { text: "¨"}
   ]
}
```

I can't remember what the logic of GetWords looks like exactly but I'm tempted to have a hard coded list of standalone diacritics and special case some logic to combine them in the same word since I'd guess they get mangled due to their y offset differing from the neighbors.

**Comment by asbjornu** [comment]
Thanks for looking into this! Doing the combination at the word level seems like a good idea. Could you perhaps use the [`StringInfo` class](https://docs.microsoft.com/en-us/dotnet/api/system.globalization.stringinfo) to perform the combination, and perhaps the [`Rune` class](https://docs.microsoft.com/en-us/dotnet/api/system.text.rune) to identify the diacritics before they are combined? There are more details [in the following article](https://docs.microsoft.com/en-us/dotnet/standard/base-types/character-encoding-introduction).

[This tweet](https://twitter.com/MrTurnerj/status/1387757133446459395) by @Turnerj also seems relevant.

**Comment by EliotJones** [comment]
So I think the issue is rather than the combining diacritics PDFs would probably contain the non-combining form. (I only looked at one file so if you're able to locate types containing the combining form I'd be very interested).

For example if we take an `o` with umlaut we could see 3 different forms:

- `ö` unicode `U+00F6` length 1. small o with umlaut
- `ö` unicode `U+006f U+0308` length 2. Combining diaeresis
- `o¨` unicode `U+006f U+00A8` length 2. Non combining diaeresis

Since PDF is a presentation rather than text interchange medium I expect all instances of accented characters will either be the 1st type, a single character with accent included in Unicode/ASCII, or the 3rd type with the diacritical mark as a standalone letter placed above the letter it is being displayed on by the rendering engine. If you are able to find instances of the combining diaeresis or other diacriticals then we may be able to use Grapheme clusters for those and it would be great to have a file for reference. My email is elioty at hotmail dot co . uk.

The set of standalone accent characters I was referring to is here https://emw3.com/unicode-accents.html

**Comment by asbjornu** [comment]
Thanks, @EliotJones. I've sent you the problematic PDF files. What's weird about the PDF that contains the problematic `ä` character, is that when I double click the word `Tjänsteinformation`, only `nsteinformation` is selected and when selecting the entire word and copying it, I get `Tjä nsteinformation`:

1. ‎0054 LATIN CAPITAL LETTER T
2. ‎006A LATIN SMALL LETTER J
3. ‎00E4 LATIN SMALL LETTER A WITH DIAERESIS
4. ‎0020 SPACE
5. ‎006E LATIN SMALL LETTER N
6. ‎0073 LATIN SMALL LETTER S
7. ‎0074 LATIN SMALL LETTER T
8. ‎0065 LATIN SMALL LETTER E
9. ‎0069 LATIN SMALL LETTER I
10. ‎006E LATIN SMALL LETTER N
11. ‎0066 LATIN SMALL LETTER F
12. ‎006F LATIN SMALL LETTER O
13. ‎0072 LATIN SMALL LETTER R
14. ‎006D LATIN SMALL LETTER M
15. ‎0061 LATIN SMALL LETTER A
16. ‎0074 LATIN SMALL LETTER T
17. ‎0069 LATIN SMALL LETTER I
18. ‎006F LATIN SMALL LETTER O
19. ‎006E LATIN SMALL LETTER N

So the character is copied (from Preview on Mac) as a combined character, but with an odd space added afterward, exactly like `ContentOrderTextExtractor.GetText()` does.

**Comment by EliotJones** [comment]
Hi @asbjornu can you try the latest nightly build when you have time please, it should fix the issue you were encountering 🤞

**Comment by asbjornu** [comment]
@EliotJones, thank you so much! The diacritic issue indeed seems to be solved. Do you have an idea what causes the Mac PDF to have tabs instead of spaces when using `ContentOrderTextExtractor.GetText(page)`?

**Comment by asbjornu** [comment]
I'll close this one and open a new issue for the tab problem.

**Comment by asbjornu** [comment]
Fixed in #440.
